File: chap12/test-code/Test4/udp_chat_server.py
# udp_chat_server.py
import socket

HOST = ""          # 모든 인터페이스
PORT = 2600        # 네가 쓰던 UDP 포트 그대로 써도 됨
BUF = 1024

# 접속 클라이언트 주소 목록 (set으로 중복 방지)
clients = set()

def main():
    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
        s.bind((HOST, PORT))
        print(f"[SERVER] UDP 채팅 서버 시작 (*:{PORT})")

        while True:
            data, addr = s.recvfrom(BUF)
            msg = data.decode("utf-8", errors="replace")

            # 새 클라이언트면 리스트에 추가
            if addr not in clients:
                clients.add(addr)
                print(f"[JOIN] {addr} 접속, 현재 인원: {len(clients)}")

            print(f"[RECV] {addr} : {msg}")

            # broadcast (보낸 애 포함/제외는 취향)
            for c in clients:
                if c != addr:  # 보낸 애 제외하고 싶으면 이 줄 유지, 포함하려면 if 지워
                    s.sendto(f"{addr}: {msg}".encode("utf-8"), c)

if __name__ == "__main__":
    main()

File: chap12/test-code/Test4/UdpChatServer.java
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.util.HashSet;
import java.util.Set;

public class UdpChatServer {
    private int port;
    private int bufSize;
    private Set<ClientAddr> clients = new HashSet<>();

    public UdpChatServer(int port, int bufSize) {
        this.port = port;
        this.bufSize = bufSize;
    }

    public void start() throws Exception {
        byte[] buf = new byte[bufSize];

        try (DatagramSocket socket = new DatagramSocket(port)) {
            System.out.println("[SERVER] Java UDP Chat Server started (*:" + port + ")");

            while (true) {
                DatagramPacket packet = new DatagramPacket(buf, buf.length);
                socket.receive(packet);

                String msg = new String(packet.getData(), 0, packet.getLength(), "UTF-8");
                ClientAddr sender = new ClientAddr(packet.getAddress(), packet.getPort());

                if (!clients.contains(sender)) {
                    clients.add(sender);
                    System.out.println("[JOIN] " + sender + " joined, users: " + clients.size());
                }

                System.out.println("[RECV] " + sender + " : " + msg);

                // broadcast message to all other clients
                String toSend = sender + " : " + msg;
                byte[] out = toSend.getBytes("UTF-8");
                for (ClientAddr c : clients) {
                    if (!c.equals(sender)) {
                        DatagramPacket outPacket =
                                new DatagramPacket(out, out.length, c.addr, c.port);
                        socket.send(outPacket);
                    }
                }
            }
        }
    }

    private static class ClientAddr {
        InetAddress addr;
        int port;

        ClientAddr(InetAddress addr, int port) {
            this.addr = addr;
            this.port = port;
        }

        @Override
        public int hashCode() {
            return addr.hashCode() * 31 + port;
        }

        @Override
        public boolean equals(Object o) {
            if (!(o instanceof ClientAddr)) return false;
            ClientAddr other = (ClientAddr) o;
            return addr.equals(other.addr) && port == other.port;
        }

        @Override
        public String toString() {
            return addr.getHostAddress() + ":" + port;
        }
    }

    public static void main(String[] args) throws Exception {
        UdpChatServer server = new UdpChatServer(2600, 1024);
        server.start();
    }
}
